File: backend/guard.ts
import { createHash } from "node:crypto";
import type { Policy, ToolRule } from "./policy";

export interface ProposedAction {
  tool: string;
  args: Record<string, unknown>;
  estCostUsd?: number;
  rationale: string;
}
export interface Decision {
  verdict: "allow" | "block";
  tool: string;
  reason: string;
  sensitivity?: ToolRule["sensitivity"];
}
export interface LedgerEntry {
  seq: number;
  ts: string;
  type: string;
  data: unknown;
  prevHash: string;
  hash: string;
}

const ZERO = "0x" + "0".repeat(64);

// In-request-path enforcement + a hash-chained, tamper-evident ledger.
// Each entry commits to the previous one, so any edit to history breaks the chain.
export class Guard {
  private rules: Map<string, ToolRule>;
  private budget: number;
  private spent = 0;
  ledger: LedgerEntry[] = [];
  private prev = ZERO;

  constructor(policy: Policy) {
    this.rules = new Map(policy.tools.map((t) => [t.tool, t]));
    this.budget = policy.budgetUsd;
  }

  private digest(seq: number, ts: string, type: string, data: unknown): string {
    return "0x" + createHash("sha256").update(this.prev + seq + ts + type + JSON.stringify(data)).digest("hex");
  }

  private append(type: string, data: unknown): LedgerEntry {
    const seq = this.ledger.length + 1;
    const ts = new Date().toISOString();
    const hash = this.digest(seq, ts, type, data);
    const entry: LedgerEntry = { seq, ts, type, data, prevHash: this.prev, hash };
    this.ledger.push(entry);
    this.prev = hash;
    return entry;
  }

  record(type: string, data: unknown): LedgerEntry {
    return this.append(type, data);
  }

  evaluate(a: ProposedAction): { decision: Decision; entry: LedgerEntry } {
    const rule = this.rules.get(a.tool);
    let verdict: "allow" | "block" = "allow";
    let reason = "Within policy";

    if (!rule) {
      verdict = "block";
      reason = `Unknown tool '${a.tool}' — not in the policy derived from the agent's code`;
    } else if (rule.sensitivity === "forbidden") {
      verdict = "block";
      reason = rule.reason;
    } else if (rule.sensitivity === "sensitive") {
      const cost = a.estCostUsd ?? 0;
      const remaining = this.budget - this.spent;
      if (cost > remaining) {
        verdict = "block";
        reason = `$${cost} exceeds remaining budget ($${remaining.toFixed(2)}) — requires human approval`;
      } else {
        this.spent += cost;
        reason = `Sensitive but within budget ($${cost} ≤ $${remaining.toFixed(2)} remaining)`;
      }
    }

    const decision: Decision = { verdict, tool: a.tool, reason, sensitivity: rule?.sensitivity };
    const entry = this.append(verdict === "allow" ? "action.allowed" : "action.blocked", {
      tool: a.tool,
      args: a.args,
      estCostUsd: a.estCostUsd ?? 0,
      rationale: a.rationale,
      decision: verdict,
      reason,
    });
    return { decision, entry };
  }

  get spentUsd(): number {
    return this.spent;
  }
  root(): string {
    return this.prev;
  }
  verifyIntegrity(): boolean {
    let prev = ZERO;
    for (const e of this.ledger) {
      const h = "0x" + createHash("sha256").update(prev + e.seq + e.ts + e.type + JSON.stringify(e.data)).digest("hex");
      if (h !== e.hash || e.prevHash !== prev) return false;
      prev = e.hash;
    }
    return true;
  }
}

File: backend/guard.py
"""In-request-path enforcement + a hash-chained, tamper-evident audit ledger."""
import hashlib
import json
from datetime import datetime, timezone

ZERO = "0x" + "0" * 64


def _iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class Guard:
    def __init__(self, policy: dict):
        self.rules = {t["tool"]: t for t in policy["tools"]}
        self.budget = float(policy["budgetUsd"])
        self.spent = 0.0
        self.ledger: list[dict] = []
        self.prev = ZERO

    def _digest(self, seq: int, ts: str, typ: str, data: dict) -> str:
        payload = self.prev + str(seq) + ts + typ + json.dumps(data, sort_keys=True, separators=(",", ":"))
        return "0x" + hashlib.sha256(payload.encode()).hexdigest()

    def record(self, typ: str, data: dict) -> dict:
        seq = len(self.ledger) + 1
        ts = _iso()
        h = self._digest(seq, ts, typ, data)
        entry = {"seq": seq, "ts": ts, "type": typ, "data": data, "prevHash": self.prev, "hash": h}
        self.ledger.append(entry)
        self.prev = h
        return entry

    def evaluate(self, tool: str, args: dict, est_cost: float = 0.0, rationale: str = ""):
        rule = self.rules.get(tool)
        verdict, reason = "allow", "Within policy"
        sensitivity = rule["sensitivity"] if rule else "unknown"

        if rule is None:
            verdict, reason = "block", f"Unknown tool '{tool}', not in the policy derived from the agent's code"
        elif rule["sensitivity"] == "forbidden":
            verdict, reason = "block", rule["reason"]
        elif rule["sensitivity"] == "sensitive":
            remaining = self.budget - self.spent
            if est_cost > remaining:
                verdict = "block"
                reason = f"${est_cost:g} exceeds remaining budget (${remaining:.2f}). Requires human approval."
            else:
                self.spent += est_cost
                reason = f"Sensitive but within budget (${est_cost:g} <= ${remaining:.2f} remaining)."

        entry = self.record(
            "action.allowed" if verdict == "allow" else "action.blocked",
            {"tool": tool, "args": args, "estCostUsd": est_cost, "rationale": rationale,
             "decision": verdict, "reason": reason},
        )
        decision = {"verdict": verdict, "tool": tool, "reason": reason, "sensitivity": sensitivity}
        return decision, entry

    @property
    def spent_usd(self) -> float:
        return self.spent

    def root(self) -> str:
        return self.prev

    def verify_integrity(self) -> bool:
        prev = ZERO
        for e in self.ledger:
            payload = prev + str(e["seq"]) + e["ts"] + e["type"] + json.dumps(e["data"], sort_keys=True, separators=(",", ":"))
            h = "0x" + hashlib.sha256(payload.encode()).hexdigest()
            if h != e["hash"] or e["prevHash"] != prev:
                return False
            prev = e["hash"]
        return True
